controllers/partition_controller.go

```diff
@@ -0,0 +1,327 @@
+				expectedPodsCount++
+			} else {
+				break
+			}
+		}
+	}
+
+	if revisionCounts[lastRevision] != expectedPodsCount {
+		log.Info("Pod count is different from the expected number", "revision", lastRevision, "expected", expectedPodsCount, "actual", revisionCounts[lastRevision])
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// isMySQLClusterHealthy checks the health status of a given MySQLCluster.
+func (r *StatefulSetPartitionReconciler) isMySQLClusterHealthy(cluster *mocov1beta2.MySQLCluster) bool {
+	return meta.IsStatusConditionTrue(cluster.Status.Conditions, mocov1beta2.ConditionHealthy)
+}
+
+// nextRolloutTargetIndex returns the index of the next rollout target Pod.
+// The index is calculated by subtracting 1 from the current partition.
+// If there is no rollout target, it returns -1.
+func (r *StatefulSetPartitionReconciler) nextRolloutTargetIndex(sts *appsv1.StatefulSet) int {
+	if sts.Spec.UpdateStrategy.RollingUpdate == nil || sts.Spec.UpdateStrategy.RollingUpdate.Partition == nil {
+		return -1
+	}
+
+	return int(*sts.Spec.UpdateStrategy.RollingUpdate.Partition) - 1
+}
+
+// getMySQLCluster retrieves the MySQLCluster release that owns a given StatefulSet.
+func (r *StatefulSetPartitionReconciler) getMySQLCluster(ctx context.Context, sts *appsv1.StatefulSet) (*mocov1beta2.MySQLCluster, error) {
+	for _, ownerRef := range sts.GetOwnerReferences() {
+		if ownerRef.Kind != "MySQLCluster" {
+			continue
+		}
+
+		cluster := &mocov1beta2.MySQLCluster{}
+		if err := r.Get(ctx, types.NamespacedName{Name: ownerRef.Name, Namespace: sts.Namespace}, cluster); err != nil {
+			return nil, err
+		}
+
+		return cluster, nil
+	}
+
+	return nil, fmt.Errorf("StatefulSet %s/%s has no owner reference to MySQLCluster", sts.Namespace, sts.Name)
+}
+
+// needPartitionUpdate returns true if the StatefulSet needs to update partition.
+func (r *StatefulSetPartitionReconciler) needPartitionUpdate(sts *appsv1.StatefulSet) bool {
+	if sts.Annotations[constants.AnnForceRollingUpdate] == "true" {
+		return false
+	}
+	if sts.Spec.UpdateStrategy.RollingUpdate == nil || sts.Spec.UpdateStrategy.RollingUpdate.Partition == nil {
+		return false
+	}
+
+	return *sts.Spec.UpdateStrategy.RollingUpdate.Partition > 0
+}
+
+// patchNewPartition patches the new partition of a StatefulSet.
+func (r *StatefulSetPartitionReconciler) patchNewPartition(ctx context.Context, sts *appsv1.StatefulSet) error {
+	oldPartition := *sts.Spec.UpdateStrategy.RollingUpdate.Partition
+	newPartition := oldPartition - 1
+
+	if oldPartition == 0 {
+		return nil
+	}
+
+	patch := client.MergeFrom(sts.DeepCopy())
+	sts.Spec.UpdateStrategy.RollingUpdate.Partition = &newPartition
+
+	if err := r.Client.Patch(ctx, sts, patch); err != nil {
+		return fmt.Errorf("failed to patch new partition to StatefulSet %s/%s: %w", sts.Namespace, sts.Name, err)
+	}
+
+	r.Recorder.Eventf(sts, corev1.EventTypeNormal, "PartitionUpdate", "Updated partition from %d to %d", oldPartition, newPartition)
```

Comment:
Please output this information as log as well.

Reply:
Thanks! added:
https://github.com/cybozu-go/moco/pull/633/commits/abecf3569bb6ed4a5bc22428f4adf100a9d6b7b4#diff-7177516db0c5ea51dd9a5d501f166358b43c80e7e9db27bff587bc357d404bc0R310